File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
import java.util.*;
class Solution {
    public int solution(String dartResult) {
        int answer = 0;
        int[] num = new int[3];
        int index = 0;
        String str = "";
        for(int i = 0; i < dartResult.length(); i++) {
            char c = dartResult.charAt(i);
            if(c >= '0' && c <= '9') { //다트 점수
                str += String.valueOf(c);
            }
            else if(c =='D' || c == 'S' || c == 'T') { //보너스
                int n = Integer.parseInt(str);
                switch(c) {
                    case 'D':
                        n = (int)Math.pow(n,2);
                        break;
                    case 'S':
                        n = (int)Math.pow(n,1);
                        break;
                    case 'T':
                        n = (int)Math.pow(n,3);
                        break;
                    default:
                        break;
                }
                num[index++] = n;
                str = "";
            } 
            else { // 옵션
                if(c == '*') {
                    if(index == 1) {
                        num[index-1] *= 2;
                    }
                    else {
                        num[index-1] *= 2;
                        num[index-2] *= 2;
                    }
                }
                else {
                    num[index-1] *= -1;
                }
                
            }
            
            
        }
        for(int n : num) {
            System.out.print(n + " ");
            answer += n;
        }
        
        return answer;
    }
}

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.js
function solution(dartResult) {
    var answer = 0;
    let str = '';
    let index = -1;
    const nums = [];
    for(let i = 0; i < dartResult.length; i++) {
        const c = dartResult[i];
        if (c >= '0' && c <= '9') {
            str += c;
        }
        else if (c === 'S' || c === 'D' || c === 'T') {
            let num = parseInt(str);
            
            index++;
            if (c === 'S') {
               num **=  1; 
            } else if (c === 'D') {
                num **= 2;
            } else num **= 3; 
            str = '';
            nums[index] = num;
        } else {
            if (c === '*') {
                nums[index] *= 2;
                if (index > 0) nums[index - 1] *= 2;
            } else {
                nums[index] *= -1
            }
        }
    }
    nums.forEach((num) => answer += num);
    return answer;
}